<!--
%\VignetteIndexEntry{Counting Hits and Enrichment/depletion Analysis for -->
<!--Repeat Elements} 
%\VignettePackage{rmskStats}
%\VignetteEngine{knitr::knitr}
-->

```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

# The rmskStats Pakcage
The rmskStats package is a standard R package providing tools for
counting gene hits and enrichment/depletion analysis for
RepeatMasker-annotated repeat elements with RNA-seq data.

## Counting Hits
Most of the reads that map to repeat elements are prone to align to
multiple postions on the genome. To avoid a read to be counted
multiple times, this package divides the count of a read by the number
of reported alignaments, the NH column from the bam file.

Let us start an example with two main functions: **sanitizeReads()** and
**summarizeOverlaps.adjNH()**. The  **sanitizeReads**_ function prepares
the queries to be counted by  **summarizeOverlaps.adjNH**. It produces
a list of GRanges of unique read with "count" (the number of copies of
the unique read) and "NH" metadata. 

The **summarizeOverlaps.adjNH** function takes the return of
**sanitizeReads()** as an input and counting the hits for the given
genomic feature with adjustment of the the NH column. 

## Example
```{r library}
suppressPackageStartupMessages(library(SGSeq))
library(rmskStats)
```

### sanitizeRead
```{r sanitizeRead}
bam_dir <- system.file("extdata", "bams", package="SGSeq")
bam_files <- list.files(bam_dir, pattern="\\.bam$", full.names=TRUE)
all.reads <- sanitizeReads(bamFiles=bam_files, cores=2, singleEnd=TRUE,
                           verbose=TRUE)
```
The GRanges element of **all.reads** arguement has to have "count" and "NH" metadata columns.

```{r all.reads}
all.reads[[1]]
```

### RMSK track from UCSC
The sub.hg19 (hg19.rmsk)  RMSK tracks is available in this package. It is a subset of RMSK of hg19 genome build, and it is downloaded via **rtracklayer** from UCSC genome browser on 03/08/2016. To build for an GRanges instance representing the RMSK track, please see "scripts/getRMSKFromUCSC.R" file.


### summarizeOverlaps.adjNH
The list-level of the feature below (repElements) is an unique repeat elements (repName). Now we are counting hits for the features of interest. We need to make sure the sequence levels are the same between the features and queries before calling the function.

```{r count1}
data(sub.hg19.rmsk)
seqlevelsStyle(sub.hg19.rmsk) <- "NCBI"
keeplevels <- intersect(seqlevels(sub.hg19.rmsk), seqlevels(all.reads[[1]]))
hg19.rmsk <- keepSeqlevels(sub.hg19.rmsk, keeplevels)
repElements <- split(sub.hg19.rmsk, sub.hg19.rmsk$repName)
se.repeat <- summarizeOverlaps.adjNH(features=repElements, cores=2,
                                     all.reads=all.reads,
                                     type="within",
                                     ignore.strand=TRUE,
                                     inter.feature=FALSE)
se.repeat									 
```
									 

## Example 2
One can use **summarizeOverlaps.adjNH** for general purpose. I found it useful for counting hits for ncRNA features.

```{r example 2}
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))
features <- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene, by="gene")
seqlevelsStyle(features) <- "NCBI"
keeplevels <- intersect(seqlevels(features), seqlevels(all.reads[[1]]))
features <- keepSeqlevels(features, keeplevels)

se <- summarizeOverlaps.adjNH(features=features, cores=2,
                              all.reads=all.reads,
                              type="within",
                              ignore.strand=TRUE,
                              inter.feature=FALSE)
se							  
```


## How to get RMSK tracks from UCSC using rtracklayer
```{r getRMSK, evel=FALSE}
library(rtracklayer)
library(GenomicRanges)
mySession = browserSession("UCSC")
genome(mySession) <- "hg38"
tbl.rmsk <- getTable(ucscTableQuery(mySession, track="rmsk", table="rmsk"))

hg38.rmsk <- GRanges(seqnames=Rle(tbl.rmsk$genoName),
                     ranges=IRanges(start=tbl.rmsk$genoStart,
                         end=tbl.rmsk$genoEnd),
                     strand=tbl.rmsk$strand)
n <- setdiff(colnames(tbl.rmsk), c("genoName", "genoStart", "genoEnd", "strand"))
mcols(hg38.rmsk) <- tbl.rmsk[, n]
hg38.rmsk
```
